poly-commitment/src/srs.rs

```diff
@@ -150,13 +145,75 @@ where
         // because the commitment to the polynomial x^i is just self.g[i], we can obtain
         // commitments to the normalized Lagrange polynomials by applying IFFT to the
         // vector self.g[0..n].
-        let mut lg: Vec<<G as AffineCurve>::Projective> =
-            self.g[0..n].iter().map(|g| g.into_projective()).collect();
-        domain.ifft_in_place(&mut lg);
+        //
+        //
+        // Further still, we can do the same trick for 'chunked' polynomials.
+        //
+        // Recall that a chunked polynomial is some f of degree k*n - 1 with
+        // f(x) = f_0(x) + x^n f_1(x) + ... + x^{(k-1) n} f_{k-1}(x)
+        // where each f_i has degree n-1.
+        //
+        // In the above, if we set u = [ 1, x^2, ... x^{n-1}, 0, 0, .., 0 ]
+        // then we effectively 'zero out' any polynomial terms higher than x^{n-1}, leaving
+        // us with the 'partial Lagrange polynomials' that contribute to f_0.
+        //
+        // Similarly, u = [ 0, 0, ..., 0, 1, x^2, ..., x^{n-1}, 0, 0, ..., 0] with n leading
+        // zeros 'zeroes out' all terms except the 'partial Lagrange polynomials' that
+        // contribute to f_1, and likewise for each f_i.
+        //
+        // By computing each of these, and recollecting the terms as a vector of polynomial
+        // commitments, we obtain a chunked commitment to the L_i polynomials.
+        let srs_size = self.g.len();
+        let num_unshifteds = (n + srs_size - 1) / srs_size;
+        let mut unshifted = Vec::with_capacity(num_unshifteds);
+
+        // For each chunk
+        for i in 0..num_unshifteds {
+            // Initialize the vector with zero curve points
+            let mut lg: Vec<<G as AffineCurve>::Projective> =
+                vec![<G as AffineCurve>::Projective::zero(); n];
+            // Overwrite the terms corresponding to that chunk with the SRS curve points
+            let start_offset = i * srs_size;
+            let num_terms = min((i + 1) * srs_size, n) - start_offset;
+            for j in 0..num_terms {
+                lg[start_offset + j] = self.g[j].into_projective()
+            }
+            // Apply the IFFT
+            domain.ifft_in_place(&mut lg);
+            <G as AffineCurve>::Projective::batch_normalization(lg.as_mut_slice());
+            // Append the 'partial Langrange polynomials' to the vector of unshifted chunks
+            unshifted.push(lg)
+        }
 
-        <G as AffineCurve>::Projective::batch_normalization(lg.as_mut_slice());
-        self.lagrange_bases
-            .insert(n, lg.iter().map(|g| g.into_affine()).collect());
+        // If the srs size does not exactly divide the domain size
+        let shifted: Option<Vec<<G as AffineCurve>::Projective>> =
+            if num_unshifteds * self.g.len() == n {
+                None
+            } else {
+                // Initialize the vector to zero
```

Comment:
Just to make sure the code obeys the theory, are there any test cases where `k` doesn't divide `n`?

In particular the edge cases of `k - n%k = 1` and `k - n%k = k - 1`